### ISSUE TITLE
disallow unused `must_use` returned by the future given to `tokio::time::timeout`

### ISSUE THREAD

**Issue by laralove143** [tokio-rs/tokio#5541]
**Is your feature request related to a problem? Please describe.**
when we pass a future to the timeout that returns a value that is `must_use`, the `unused_must_use` lint isn't forced

this is most common in passing fallible futures, the error returned by it is just silently ignored, since propagation (`?`) inside the async block doesn't return from the entire function
```rust
if tokio::time::timeout(Duration::from_secs(60), async {
  fallible().await?;
  Ok(())
}).await.is_err() {
  handle_timeout().await;
}
```
ignoring a result would otherwise emit the warning `unused_must_use`

**Describe the solution you'd like**
im not sure actually

**Describe alternatives you've considered**
using `select! {}` with the future in one branch and `tokio::time::sleep()` in the other, but this might be less preferable to `timeout`

**Additional context**
this has happened to me, even though it was my mistake, the point of rust's safety system is so that it's harder to make mistakes

**Comment by Darksonn** [comment]
The reason that the warning is ignored here is due to the use of `is_err()`. I don't think this has anything to do with `timeout`, and I don't think there's anything we can do about it. If you ignore the timeout's error, you _will_ get a warning.

**Comment by laralove143** [comment]
yes but not ignoring the timeout's result, and ignoring the wrapped `Ok`, even when it's `#[must_use]` is allowed, i guess this is the same as `Result<Result<(), _>, _>` so maybe its expected that we handle it ourselves, but i think nested results arent common and timeout enforces that
